src/IO/WriteBufferFromS3.cpp

```diff
@@ -140,27 +142,36 @@ void WriteBufferFromS3::writePart(const String & data)
         LOG_WARNING(log, "Maximum part number in S3 protocol has reached (too many parts). Server may not accept this whole upload.");
     }
 
-    Aws::S3::Model::UploadPartRequest req;
+    size_t part_index = part_tags.size();
+    auto part_tag = std::make_shared<String>();
+    part_tags.emplace_back(part_tag);
 
-    req.SetBucket(bucket);
-    req.SetKey(key);
-    req.SetPartNumber(part_tags.size() + 1);
-    req.SetUploadId(upload_id);
-    req.SetContentLength(data.size());
-    req.SetBody(std::make_shared<Aws::StringStream>(data));
+    auto write_job = [&, part_index, part_tag, data] /// FIXME! data shall be moved properly
+    {
+        Aws::S3::Model::UploadPartRequest req;
 
-    auto outcome = client_ptr->UploadPart(req);
+        req.SetBucket(bucket);
+        req.SetKey(key);
+        req.SetPartNumber(part_index + 1);
+        req.SetUploadId(upload_id);
+        req.SetContentLength(data.size());
+        req.SetBody(std::make_shared<Aws::StringStream>(data));
 
-    LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Data size: {}", bucket, key, upload_id, data.size());
+        auto outcome = client_ptr->UploadPart(req);
 
-    if (outcome.IsSuccess())
-    {
-        auto etag = outcome.GetResult().GetETag();
-        part_tags.push_back(etag);
-        LOG_DEBUG(log, "Writing part finished. Total parts: {}, Upload_id: {}, Etag: {}", part_tags.size(), upload_id, etag);
-    }
-    else
-        throw Exception(outcome.GetError().GetMessage(), ErrorCodes::S3_ERROR);
+        LOG_TRACE(log, "Writing part. Bucket: {}, Key: {}, Upload_id: {}, Data size: {}", bucket, key, upload_id, data.size());
+
+        if (outcome.IsSuccess())
+        {
+            auto etag = outcome.GetResult().GetETag();
+            part_tag->assign(std::move(etag));
+            LOG_DEBUG(log, "Writing part finished. Upload_id: {}, Etag: {}", upload_id, part_tag);
+        }
+        else
+            throw Exception(outcome.GetError().GetMessage(), ErrorCodes::S3_ERROR);
+    };
+
+    writing_thread_pool.scheduleOrThrowOnError(write_job);
```

Comment:
Those threads will not be accounted for the query, like here https://github.com/ClickHouse/ClickHouse/pull/14150#discussion_r534413602